utils/run_utest.py

```diff
@@ -0,0 +1,617 @@
+#!/usr/bin/env python3
+"""Test script for running all DAOS unit tests
+"""
+# pylint: disable=broad-except
+import os
+import sys
+import json
+import argparse
+import shutil
+import re
+import subprocess  # nosec
+import tempfile
+import traceback
+from junit_xml import TestSuite, TestCase
+import yaml
+
+
+def write_xml_result(name, suite_junit):
+    """Write an junit result"""
+    with open(f"test_results/test_{name}.xml", "w", encoding='UTF-8') as file:
+        TestSuite.to_file(file, [suite_junit], prettyprint=True)
+
+
+def junit_name(memcheck):
+    """Get name of junit test"""
+    typestr = "native"
+    if memcheck:
+        typestr = "memcheck"
+    return f"run_utest.py.{typestr}"
+
+
+def setup_junit(memcheck):
+    """Setup single junit for whole suite"""
+    name = junit_name(memcheck)
+    test = TestCase("run_utest", name)
+    suite = TestSuite(name, [test])
+    return (suite, test)
+
+
+class BaseResults():
+    """Keep track of test results"""
+    def __init__(self):
+        """Initializes the values"""
+        self.results = {"tests": 0, "errors": 0, "failures": 0, "fail_msg": "", "error_msg": ""}
+
+    def has_failures(self):
+        """Return true if there are test failures or errors"""
+        return self.results["errors"] or self.results["failures"]
```

Comment:
For CI, we want test failures due to DAOS bugs to have this return 0, and to return 1 for failures that are not DAOS bugs.
DAOS bugs will be detected in the Junit results, and will mark the stage to be UNSTABLE in CI.
Bugs in the test code itself or CI infrastructure should mark the stage FAILED in CI.

Reply:
@JohnMalmberg That should be achieved with https://github.com/daos-stack/daos/blob/fbba631d877242701bc7ac79442d80c8b407ecc9/utils/run_utest.py#L609-L610.

If this python script fails in any way, it will exit !0, giving us the FAILURE status in Jenkins.

Reply:
Yeah, I consider anything that I've trapped to be a DAOS bug and anything else to be CI (or a bug in the script).  The args.no_fail_on_error Brian points out should prevent it from failing on issues we've trapped.   There may need to be some refinement later but this definitely shows orange on valgrind errors